### DISCUSSION TITLE
tutorials/best_practices/project_organization

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#205] [User-contributed notes]
# tutorials/best_practices/project_organization

Introduction: Since Godot has no restrictions on project structure or filesystem usage, organizing files when learning the engine can seem challenging. This tutorial suggests a workflow which shoul...

https://docs.godotengine.org/en/stable/tutorials/best_practices/project_organization.html

**Comment by RedpingDev** [comment]
Do You have best case!? I wonder about that.

**Reply by Calinou** [reply]
What do you mean by "best case"?

**Reply by RedpingDev** [reply]
Ah~ I was wondering if you had an example of actually splitting a folder into good and bad cases, like the example you posted in the code style. I've already got my question answered, but I think it would be more informative to write a document comparing the cases. good luck!

**Reply by sjorskoot** [reply]
I thought this was a pun on case sensitivity

**Comment by LethargicDeveloper** [comment]
In the Project Settings menu the default Directory Naming Convention is kabab-case, yet here, the recommendation is to use snake_case for directories. Is there a reason for this? Is the project setting default outdated or wrong?

**Reply by Calinou** [reply]
The **Directory Naming Convention** setting is only used to define project folders when creating a project in the project manager. It's not used anywhere else in the editor, so the recommendation to use snake_case within the project still applies.

The reason this is done is that it's a common convention for Git repository names to use kebab-case, but snake_case is perfectly valid too.

**Reply by Wiggleslost** [reply]
I have a lot learning to do with the cases! Do the files actually go into a folder, have not gotten really to far, one file was named Nacirema when attempted to fit all the files in the folder, will a lowercase be a possible good generator for files going into different folders made?

**Comment by douglaslassance** [comment]
Would be nice if the content of `.gdignore` was not ignored.

**Reply by Calinou** [reply]
This is being tracked in https://github.com/godotengine/godot-proposals/issues/2798.

**Comment by jauntyjocularjay** [comment]
It looks like adding the .gdignore file to a folder is not as easy as it seems. I use the same folder structure for my dabblings with Unity. I have placed my Unity files in their own folder with .gdignore, but Godot still tries to build the CS files contained within regardless.

**Reply by Calinou** [reply]
`.cs` files are not imported resources. They are also handled by the .NET toolchain anyway, which has no knowledge of `.gdignore` files and other Godot specifics. Therefore, it's expected that `.gdignore` files don't affect whether `.cs` files are compiled.

I believe you should be able to modify the `.sln` file to exclude the folder containing your Unity project. This way, it should avoid recompiling it.

**Comment by Wiggleslost** [comment]
I am new to the program, what does case sensitive mean? Is all letters supposed to be lowercase?

**Reply by Neo-Ciber94** [reply]
Case sensitive means: "HELLO" and "hello" will be considered different.

But when case insensitive: "HELLO" and "hello" will be considered the same

**Reply by Wiggleslost** [reply]
I am really confused, okay, so case insensitive like the caps lock off on a device? I know that caps lock produces the uppercase LIKE THE CAPS LOCK IS ON CURRENTLY, how would "hello" be available with case insensitive? I can press the shift button to produce an uppercase, interested in the case insensitive for the language in GDScript, I saw a camel case, is this how "GDS" came about for GDScript kind of like a camel rising up a letter from their neck for producing an uppercase letter?

**Reply by cheetahsnail** [reply]
Maybe I'm misreading your question here, but to expand on the first answer, "case sensitivity" is something we use all the time for passwords.

Lets say your password is "PassWord1234"

If the system you're trying to login to is case sensitive, then it will only accept "PassWord1234" for your login.

If the system is NOT case sensitive (AKA case insensitive), then not only will it accept "PassWord1234" but it could accept "password1234", "PASSword1234", "PaSsWoRd1234" or any combination of caps/lowercase (this would make for really bad security).

So in terms of file naming, lets say you have the asset "flower". You're working in a case insensitive environment and you've referred to it in your code as "flower" "Flower" and "FLOWER". This has caused 0 problems for you because the script is ignoring if the letters are caps/lowercase (case insensitive).

But in a case sensitive environment, your script is going to treat the "flower" "Flower" and "FLOWER" names as three different files to look for. This will cause a lot of problems because you only have one file: "flower". 

What the documentation is saying is that you need to be mindful of your naming conventions to avoid problems in the future.

**Comment by Wiggleslost** [comment]
So case insensitive is "FLOWER"? I kind of see, thanks

**Reply by sjorskoot** [reply]
No. "FLOWER" is an example of a case type (see https://danielmiessler.com/blog/a-list-of-different-case-types for more examples). Case (in)sensitivity is a property of the environment, such as Windows vs Linux. Linux cares about the casing of your files, Windows does not. What that means for you is that when your file is called "FLOWER.png", you need to spell it exactly the same when you refer to it in code. It is recommended you follow a convention so that you don't accidentally use an uppercase somewhere where it should have been a lowercase or vice versa.
